src/bin/compute-gop.cc

```diff
@@ -0,0 +1,228 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+   This code computes Goodness of Pronunciation (GOP) and extracts phone-level
+   pronunciation feature for mispronunciations detection tasks, the reference:
+
+   "Improved mispronunciation detection with deep neural network trained acoustic
+   models and transfer learning based logistic regression classifiers"
+   by Hu et al., Speech Comunication, 2015.
+
+   GOP is widely used to detect mispronunciations. The DNN-based GOP was defined
+   as the log phone posterior ratio between the canonical phone and the one with
+   the highest score.
+
+   To compute GOP, we need to compute Log Phone Posterior (LPP):
+     LPP(p) = \log p(p|\mathbf o; t_s,t_e)
+   where {\mathbf o} is the input observations, p is the canonical phone,
+   {t_s, t_e} are the start and end frame indexes.
+
+   LPP could be calculated as the average of the frame-level LPP, i.e. p(p|o_t):
+     LPP(p) = \frac{1}{t_e-t_s+1} \sum_{t=t_s}^{t_e}\log p(p|o_t)
+     p(p|o_t) = \sum_{s \in p} p(s|o_t)
+   where s is the senone label, {s|s \in p} is the states belonging to those
+   triphones whose current phone is p.
+
+   GOP is extracted from LPP:
+     GOP(p) = \log \frac{LPP(p)}{\max_{q\in Q} LPP(q)}
+
+   An array of a phone-level feature for each phone is extracted as well, which
+   could be used to train a classifier to detect mispronunciations. Normally the
+   classifier-based approach archives better performance than the GOP-based approach.
+
+   The phone-level feature is defined as:
+     {[LPP(p_1),\cdots,LPP(p_M), LPR(p_1|pi), \cdots, LPR(p_j|p_i),\cdots]}^T
+
+   where the Log Posterior Ratio (LPR) between phone p_j and p_i is defined as:
+     LPR(p_j|p_i) = \log p(p_j|\mathbf o; t_s, t_e) - \log p(p_i|\mathbf o; t_s, t_e)
+ */
+
+#include <regex>
+#include "base/kaldi-common.h"
+#include "util/common-utils.h"
+#include "hmm/transition-model.h"
+#include "hmm/hmm-utils.h"
+#include "hmm/tree-accu.h"
+#include "hmm/posterior.h"
+
+namespace kaldi {
+
+/** FrameLevelLpp compute a log posterior for pure-phones by sum the posterior
+    of the states belonging to those triphones whose current phone is the canonical
+    phone:
+
+    p(p|o_t) = \sum_{s \in p} p(s|o_t),
+
+    where s is the senone label, {s|s \in p} is the states belonging to those
+    riphones whose current phone is the canonical phone p.
+
+ */
+Vector<BaseFloat> FrameLevelLpp(SubVector<BaseFloat> prob_row,
```

Comment:
Please pass in the SubVector as a const reference, and rather than returning a Vector<BaseFloat>, make it a pointer argument.
References should be const, anything that can be const should be const.

Reply:
I'll modify it as your suggestion :)